### PR TITLE
Replace byte(A, X) by zero if A >= 32

### LINKED ISSUE #6174: Add optimisation rule: byte(x, a) where a >= 32 -> 0
From #6107

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/6174

### CODE REVIEW

**Comment by chriseth** [comment]
@bshastry this is a new optimizer rule to be fuzzed.

File: Changelog.md [hunk 1]
```diff
@@ -5,6 +5,7 @@ Language Features:
 
 
 Compiler Features:
+ * Optimizer: Add rule that replaces the BYTE opcode by 0 if the byte number is larger than 31.
```
Comment (bshastry) [inline Changelog.md]: `byte index` instead of `byte number`?

**Review by bshastry — COMMENTED** [review]
Just a minor comment regarding changelog entry

File: Changelog.md (hunk 1 above)
Comment (chriseth) [reply Changelog.md]: or just "first argument"?

Comment (bshastry) [reply Changelog.md]: First argument is good.

**Comment by leonardoalt** [comment]
Still missing the proof, I'll add it today.

**Comment by leonardoalt** [comment]
The yellow paper already defines this as 0 if `A >= 32`, so I think this doesn't need a proof.

**Comment by chriseth** [comment]
Still, we could have swapped the order of parameters or made an off-by-one error with the constant.

**Comment by leonardoalt** [comment]
If we implement `BYTE` exactly as the yellow paper defines it then the proof of the rule is trivial.

**Comment by bshastry** [comment]
Minor update: No negative results from fuzzing so far (about 15 million iterations).

**Review by ekpyron — APPROVED** [review]
Looks good to me.

**Pull request merged**